File: RRCPClient.java

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.concurrent.locks.Lock;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Alex
 */
public class RRCPClient {

    private Socket socket;
    private DataInputStream dataInput;
    private DataOutputStream dataOutput;
    private String hostIP;
    private int hostPort;
    private int packetTimeOut;
    private volatile boolean connected = false;
    private boolean connecting = false;
    private Thread heartBeatThread;
    private PacketHandler packetHandler;
    private int heartBeatDelay = 0;
    private volatile byte currentAddress = -1;
    private final int TIMEOUT_NUM = 5;
    private boolean autoReconnect = true;

    private static enum PacketTypes {

        Byte((byte) 1),
        Integer((byte) 2),
        Double((byte) 3),
        Long((byte) 4),
        Short((byte) 5),
        Float((byte) 6),
        Command((byte) 7),
        String((byte) 8),
        Boolean((byte) 9),
        HeartBeat((byte) 21),
        ClientCommand((byte) 30),
        ClientCommandDouble((byte) 31),
        ClientCommandDoubleArray((byte) 32),
        DoubleArray((byte) 10),
        ByteArray((byte) 11),
        IntegerArray((byte) 12),
        LongArray((byte) 13),
        ShortArray((byte) 14),
        FloatArray((byte) 15);

        private final byte id;

        private PacketTypes(byte b) {
            this.id = b;
        }

        public byte getID() {
            return id;
        }
    }

    public RRCPClient(String host, int port, int timeout, boolean autoRe) {
        this.hostIP = host;
        this.hostPort = port;
        this.packetTimeOut = timeout / TIMEOUT_NUM;
        this.autoReconnect = autoRe;
    }

    public synchronized void connect() {
        if (isConnected()) {

        } else if (isConnecting()) {

        } else {
            try {
                connecting = true;
                socket = new Socket(hostIP, hostPort);
                dataInput = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
                dataOutput = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
                packetHandler = new PacketHandler();
                heartBeatThread = new Thread(new HeartBeatThread());
                connected = true;
                heartBeatThread.start();
            } catch (IOException ex) {
                Logger.getLogger(RRCPClient.class.getName()).log(Level.SEVERE, null, ex);
            } finally {
                connecting = false;
            }

        }
    }

    public boolean isConnected() {
        return connected;
    }

    public boolean isConnecting() {
        return connecting;
    }

    public byte getCurrentAddress() {
        return currentAddress;
    }

    private synchronized byte addCurrentAddress() {
        if (currentAddress == 100) {
            currentAddress = -1; //if last packet sent had address of 50 change currentAdress back to -1 to be changed to 0
        }
        packetHandler.packetQueue[++currentAddress] = null;
        return getCurrentAddress();
    }

    public byte sendCommand(String command) {
        byte address;
        if (isConnected()) {
            try {
                address = addCurrentAddress();
                dataOutput.write(PacketTypes.Command.getID());
                addAddressCommand(address, command);
                dataOutput.flush();
                return address;
            } catch (IOException ex) {

            }
        } else {

        }
        return -2;
    }

    private void addAddressCommand(byte address, String command) throws IOException {
        dataOutput.write(address);
        dataOutput.writeUTF(command);
    }

    public byte sendCommandWithNumber(String command, Number n) {
        if (isConnected()) {
            try {
                byte address = addCurrentAddress();
                if (n instanceof Integer) {
                    dataOutput.write(PacketTypes.Integer.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt((int) n);
                    dataOutput.flush();
                    return address;
                } else if (n instanceof Double) {
                    dataOutput.write(PacketTypes.Double.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeDouble((double) n);
                    dataOutput.flush();
                    return address;
                } else if (n instanceof Short) {
                    dataOutput.write(PacketTypes.Short.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeDouble((short) n);
                    dataOutput.flush();
                    return address;
                } else if (n instanceof Byte) {
                    dataOutput.write(PacketTypes.Byte.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeDouble((byte) n);
                    dataOutput.flush();
                    return address;
                } else if (n instanceof Long) {
                    dataOutput.write(PacketTypes.Long.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeDouble((long) n);
                    dataOutput.flush();
                    return address;
                } else if (n instanceof Float) {
                    dataOutput.write(PacketTypes.Float.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeDouble((float) n);
                    dataOutput.flush();
                    return address;
                } else {
                    System.err.println("Really... I don't what to support that number type...");
                    return -2;
                }
            } catch (IOException ex) {
                return -2;
            }
        } else {
            return -2;
        }
    }

    public byte sendCommandWithNumberArray(String command, Object array) {
        if (isConnected()) {
            try {
                byte address = addCurrentAddress();
                if (array instanceof int[]) {
                    dataOutput.write(PacketTypes.IntegerArray.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt(((int[]) array).length);
                    for (int i = 0; i < ((int[]) array).length; i++) {
                        dataOutput.writeInt(((int[]) array)[i]);
                    }
                    dataOutput.flush();
                    return address;
                } else if (array instanceof double[]) {
                    dataOutput.write(PacketTypes.DoubleArray.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt(((double[]) array).length);
                    for (int i = 0; i < ((double[]) array).length; i++) {
                        dataOutput.writeDouble(((double[]) array)[i]);
                    }
                    dataOutput.flush();
                    return address;
                } else if (array instanceof short[]) {
                    dataOutput.write(PacketTypes.ShortArray.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt(((short[]) array).length);
                    for (int i = 0; i < ((short[]) array).length; i++) {
                        dataOutput.writeShort(((short[]) array)[i]);
                    }
                    dataOutput.flush();
                    return address;
                } else if (array instanceof byte[]) {
                    dataOutput.write(PacketTypes.ByteArray.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt(((byte[]) array).length);
                    for (int i = 0; i < ((byte[]) array).length; i++) {
                        dataOutput.write(((byte[]) array)[i]);
                    }
                    dataOutput.flush();
                    return address;
                } else if (array instanceof long[]) {
                    dataOutput.write(PacketTypes.LongArray.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt(((long[]) array).length);
                    for (int i = 0; i < ((long[]) array).length; i++) {
                        dataOutput.writeLong(((long[]) array)[i]);
                    }
                    dataOutput.flush();
                    return address;
                } else if (array instanceof float[]) {
                    dataOutput.write(PacketTypes.FloatArray.getID());
                    addAddressCommand(address, command);
                    dataOutput.writeInt(((float[]) array).length);
                    for (int i = 0; i < ((float[]) array).length; i++) {
                        dataOutput.writeFloat(((float[]) array)[i]);
                    }
                    dataOutput.flush();
                    return address;
                } else {
                    System.err.println("This doesn't support that number array");
                    return -2;
                }
            } catch (IOException ex) {
                return -2;
            }
        } else {
            return -2;
        }
    }

    public byte sendCommandWithBoolean(String command, boolean b) {
        byte address;
        if (isConnected()) {
            try {
                address = addCurrentAddress();
                dataOutput.write(PacketTypes.Boolean.getID());
                addAddressCommand(address, command);
                dataOutput.writeBoolean(b);
                dataOutput.flush();
                return address;
            } catch (IOException ex) {

            }
        } else {

        }
        return -2;
    }

    public byte sendCommandWithString(String command, String s) {
        byte address;
        if (isConnected()) {
            try {
                address = addCurrentAddress();
                dataOutput.write(PacketTypes.String.getID());
                addAddressCommand(address, command);
                dataOutput.writeUTF(s);
                dataOutput.flush();
                return address;
            } catch (IOException ex) {

            }
        } else {

        }
        return -2;
    }

    private class PacketHandler implements Runnable {
        
        private Packet[] packetQueue;
        private Packet beatQueue;
        private Thread mainThread;
        
        public PacketHandler() {
            this.packetQueue = new Packet[101]; //Makes packet queue
            this.beatQueue = null;
            this.mainThread = new Thread(this);
            this.mainThread.start();
        }
        
        public void run() {
            while (isConnected()) {
                try {
                    while (dataInput.available() > 0) { //Reads packet when it comes in from input stream
                        new Packet((byte)readNumber(PacketTypes.Byte.getID())); //Makes new packet of id read from byte
                    }
                } catch (IOException ex) {
                    
                }
                try {
                    Thread.sleep(TIMEOUT_NUM);
                } catch (InterruptedException ex) {
                    
                }
            }
        }
        
         private Packet getHeartBeat() {
            int i = 0;
            while (beatQueue == null) {
                if (i > packetTimeOut) {
                    System.err.println("SERVER DID NOT RESPOND!!!");
                    return new Packet((byte) 100); //Returns error packet
                }
                try {
                    Thread.sleep(TIMEOUT_NUM);
                } catch (InterruptedException ex) {
                    
                }
                ++i;
            }
            setHeartBeatDelay(i);
            Packet p = beatQueue;
            resetBeatQueue();
            return p;
        }

        private Packet getPacket(byte address) {
            int i = 0;
            while (packetQueue[address] == null) {
                ++i;
                if (i > packetTimeOut + 10) {
                    System.err.println("Could not find packet! Packet lost!");
                    return null;
                }
                try {
                    Thread.sleep(TIMEOUT_NUM);
                } catch (InterruptedException ex) {

                }
            }
            Packet p = packetQueue[address];
            packetQueue[address] = null;
            return p;
        }

        public void addPacketToBeatQueue(Packet p) {
            beatQueue = p;
        }

        public void addPacketToQueue(Packet p) {
            packetQueue[p.address] = p;
        }

        public void resetBeatQueue() {
            beatQueue = null;
        }

    }
    
    public Object readNumberPacket(byte address) {
        if(isConnected()) {
            Packet isNull = packetHandler.getPacket(address);
            if(isNull == null) {
                return null;
            }
            return isNull.getData();
        }
        return null;
    }
    
    private class Packet {

        private byte id;
        public Object data;
        private byte address;

        public Packet(byte id) {
            this.id = id;
            if (id == PacketTypes.HeartBeat.getID()) { //Determains what type of packet it is from id
                packetHandler.addPacketToBeatQueue(this);
            } else if (id <= PacketTypes.Float.getID()) { //Numbers
                address = (byte)readNumber(PacketTypes.Byte.getID());
                data = readNumber(id);
                packetHandler.addPacketToQueue(this);
            } else if (id == PacketTypes.Boolean.getID()) {
                address = (byte)readNumber(PacketTypes.Byte.getID());
                data = readBoolean();
                packetHandler.addPacketToQueue(this);
            } else if (id == PacketTypes.String.getID()) {
                address = (byte)readNumber(PacketTypes.Byte.getID());
                data = readString();
                packetHandler.addPacketToQueue(this);
            } else if (id >= PacketTypes.DoubleArray.getID() && id <= PacketTypes.FloatArray.getID()) {
                address = (byte)readNumber(PacketTypes.Byte.getID());
                data = readNumberArray(id);
                packetHandler.addPacketToQueue(this);
            } else if (id == 100) { //Error packet
            } else {
                data = null;
                System.err.println("Packet not reconized!!!"); //Unknow packet id
            }
        }

        public Object getData() {
            return data;
        }

        public byte getID() {
            return id;
        }
    }
    
    private Number readNumber(byte id) {
        try {
            if (id == PacketTypes.Byte.getID()|| id == PacketTypes.ByteArray.getID()) return dataInput.readByte();
            else if (id == PacketTypes.Integer.getID() || id == PacketTypes.IntegerArray.getID()) return dataInput.readInt();
            else if (id == PacketTypes.Double.getID() || id == PacketTypes.DoubleArray.getID()) return dataInput.readDouble();
            else if (id == PacketTypes.Long.getID() || id == PacketTypes.LongArray.getID()) return dataInput.readLong();
            else if (id == PacketTypes.Short.getID() || id == PacketTypes.ShortArray.getID()) return dataInput.readShort();
            else if (id == PacketTypes.Float.getID() || id == PacketTypes.FloatArray.getID()) return dataInput.readFloat();
            else return -1;
        } catch (IOException ex) {
            
        }
        return -1;
    }
    
    private Object readNumberArray(byte id) {
        int length = (int)readNumber(PacketTypes.Integer.getID());
        Number[] n = new Number[length];
        for (int i = 0; i < length; i++) {
            n[i] = readNumber(id);
        }
        return n;
    }
    
    private boolean readBoolean() {
        try {
            return dataInput.readBoolean();
        } catch (IOException ex) {
            
        }
        return false; 
    }
    
    private String readString() {
        try {
            return dataInput.readUTF();
        } catch (IOException ex) {
            
        }
        return null;
    }

    private void close() {
        try {
            this.connected = false;
            if (socket != null) {
                this.heartBeatThread.interrupt();
                this.socket.close();
            }
        } catch (IOException ex) {
            
        } finally {
            if(autoReconnect) {
                this.connect();
            }
        }
    }
    
    public void disconnect() {
        if(autoReconnect) {
            this.autoReconnect = false;
            this.close();
            this.autoReconnect = true;
        } else this.close();
    }
    //Heartbeat stuff

    private class HeartBeatThread implements Runnable {

        @Override
        public void run() {
            while (isConnected() && !Thread.currentThread().isInterrupted()) {
                if (sendHeartBeat()) {
                    Thread.sleep(250);
                } else {
                    Thread.currentThread().interrupt();
                    close();
                    break;
                }
            }
        }

        private boolean sendHeartBeat() {
            this.sendHeartBeatCommand();
            if (packetHandler.getHeartBeat().getID() == 21) {
                connected = true;
                return true;
            } else {
                return false;
            }
        }

        private void sendHeartBeatCommand() {
            if (isConnected()) {
                try {
                    dataOutput.write(PacketTypes.HeartBeat.getID());
                    dataOutput.flush();
                } catch (IOException ex) {

                }
            } else {

            }
        }
    }

    private void setHeartBeatDelay(int i) {
        this.heartBeatDelay = i;
    }

    public int getDelay() {
        return TIMEOUT_NUM * this.heartBeatDelay;
    }
}

File: RRCPClient.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace RRCP
{
    public class RRCPClient
    {

        private System.Net.Sockets.TcpClient socket;
        private string hostIp;
        private int timeout;
        private int hostPort;
        private volatile bool connected = false, connecting = false;
        private BinaryReader dataInput;
        private BinaryWriter dataOutput;
        private Object connectLock = new Object(), outputLock = new Object(), addressLock = new Object();
        private Thread heartBeatThread;
        private int heartBeatDelay = 0;
        private volatile sbyte currentAddress = -1;
        private const int TIMEOUT_NUM = 5;
        private Boolean autoReconnect = false;
        public PacketHandler packetHandler;

        public enum PacketTypes : sbyte
        {
            Byte = 1,
            Integer,
            Double,
            Long,
            Short,
            Float,
            Command,
            String,
            Boolean,
            DoubleArray,
            ByteArray,
            IntegerArray,
            LongArray,
            ShortArray,
            FloatArray,
            HeartBeat = 21,
            ClientCommand = 30,
            ClientCommandDouble,
            ClientCommandDoubleArray
        }

        public RRCPClient(string host, int timeout, int port)
        {
            this.hostIp = host;
            this.hostPort = port;
            this.timeout = timeout;
        }

        public void Connect()
        {
            lock (connectLock)
            {
                if (isConnected() || isConnecting())
                {

                }
                else
                {
                    try
                    {
                        connecting = true;
                        socket = new System.Net.Sockets.TcpClient(hostIp, hostPort);
                        dataInput = new BinaryReader(socket.GetStream(), System.Text.Encoding.UTF8);
                        dataOutput = new BinaryWriter(socket.GetStream(), System.Text.Encoding.UTF8);
                        packetHandler = new PacketHandler(this);
                        heartBeatThread = new Thread(HeartBeatRun);
                        heartBeatThread.Start();
                        connected = true;
                    }
                    catch (Exception e)
                    {
                        connected = false;
                    }
                    finally
                    {
                        connecting = false;
                    }
                }
            }
        }

        public void Dissconnect()
        {
            if (isConnected())
            {
                dataOutput.Close();
                dataInput.Close();
                socket.Close();
            }
        }
        public bool isConnected()
        {
            return connected;
        }

        public bool isConnecting()
        {
            return connecting;
        }

        public sbyte getCurrentAddress()
        {
            return currentAddress;
        }

        private sbyte addCurrentAddress()
        {
            lock (addressLock)
            {
                if (currentAddress == 100)
                {
                    currentAddress = -1; //if last packet sent had address of 50 change currentAdress back to -1 to be changed to 0
                }
                packetHandler.packetQueue[++currentAddress] = null;
                return getCurrentAddress();
            }
        }
        private void addAddressCommand(sbyte address, string command)
        {
            dataOutput.Write(address);
            byte[] b = BitConverter.GetBytes((ushort)command.Length);
            Array.Reverse(b);
            dataOutput.Write(b);
            foreach(char c in command) 
            {
                dataOutput.Write(c);
            }
        }
        public sbyte sendCommand(string command)
        {
            if (isConnected())
            {
                sbyte address;
                lock (outputLock)
                {
                    address = addCurrentAddress();
                    dataOutput.Write((sbyte)PacketTypes.Command);
                    addAddressCommand(address, command);
                    dataOutput.Flush();
                }
                return address;
            }
            else
            {

            }
            return -2;
        }
        public sbyte sendCommandWithNumber(string command, ValueType v)
        {
            if (isConnected())
            {
                sbyte address = addCurrentAddress();
                if (v is int)
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.Integer);
                        addAddressCommand(address, command);
                        dataOutput.Write((int)v);
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (v is double)
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.Double);
                        addAddressCommand(address, command);
                        byte[] b = BitConverter.GetBytes((double)v);
                        Array.Reverse(b);
                        dataOutput.Write(b);
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (v is short)
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.Short);
                        addAddressCommand(address, command);
                        dataOutput.Write((short)v);
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (v is sbyte)
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.Byte);
                        addAddressCommand(address, command);
                        dataOutput.Write((sbyte)v);
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (v is long)
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.Long);
                        addAddressCommand(address, command);
                        dataOutput.Write((long)v);
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (v is float)
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.Float);
                        addAddressCommand(address, command);
                        dataOutput.Write((float)v);
                        dataOutput.Flush();
                    }
                    return address;
                }
                else
                {
                    return -2;
                }
            }
            else
            {
                return -2;
            }
        }

        public sbyte sendCommandWithNumberArray(string command, Object array)
        {
            if (isConnected())
            {
                sbyte address = addCurrentAddress();
                if (array is int[])
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.IntegerArray);
                        addAddressCommand(address, command);
                        dataOutput.Write(((int[])array).Length);
                        for (int i = 0; i < ((int[])array).Length; i++)
                        {
                            dataOutput.Write(((int[])array)[i]);
                        }
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (array is double[])
                {
                    Debug.WriteLine("Writing array");
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.DoubleArray);
                        addAddressCommand(address, command);
                        //dataOutput.Write(((double[])array).Length);
                        byte[] l = BitConverter.GetBytes(((double[])array).Length);
                        Array.Reverse(l);
                        dataOutput.Write(l);
                        for (int i = 0; i < ((double[])array).Length; i++)
                        {
                            byte[] b = BitConverter.GetBytes(((double[])array)[i]);
                            Array.Reverse(b);
                            dataOutput.Write(b);
                        }
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (array is short[])
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.ShortArray);
                        addAddressCommand(address, command);
                        dataOutput.Write(((short[])array).Length);
                        for (int i = 0; i < ((short[])array).Length; i++)
                        {
                            dataOutput.Write(((short[])array)[i]);
                        }
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (array is sbyte[])
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.ByteArray);
                        addAddressCommand(address, command);
                        dataOutput.Write(((byte[])array).Length);
                        for (int i = 0; i < ((sbyte[])array).Length; i++)
                        {
                            dataOutput.Write(((byte[])array)[i]);
                        }
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (array is long[])
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.LongArray);
                        addAddressCommand(address, command);
                        dataOutput.Write(((long[])array).Length);
                        for (int i = 0; i < ((long[])array).Length; i++)
                        {
                            dataOutput.Write(((long[])array)[i]);
                        }
                        dataOutput.Flush();
                    }
                    return address;
                }
                else if (array is float[])
                {
                    lock (outputLock)
                    {
                        dataOutput.Write((sbyte)PacketTypes.FloatArray);
                        addAddressCommand(address, command);
                        dataOutput.Write(((float[])array).Length);
                        for (int i = 0; i < ((float[])array).Length; i++)
                        {
                            dataOutput.Write(((float[])array)[i]);
                        }
                        dataOutput.Flush();
                    }
                    return address;
                }
                else
                {
                    return -2;
                }
            }
            else
            {
                return -2;
            }
        }

        public sbyte sendCommandWithBoolean(string command, bool b)
        {
            sbyte address;
            if (isConnected())
            {

                lock (outputLock)
                {
                    address = addCurrentAddress();
                    dataOutput.Write((sbyte)PacketTypes.Boolean);
                    addAddressCommand(address, command);
                    dataOutput.Write(b);
                    dataOutput.Flush();
                }
                return address;

            }
            else
            {

            }
            return -2;
        }

        public sbyte sendCommandWithString(string command, string s)
        {
            sbyte address;
            if (isConnected())
            {

                lock (outputLock)
                {
                    address = addCurrentAddress();
                    dataOutput.Write((sbyte)PacketTypes.String);
                    addAddressCommand(address, command);
                    dataOutput.Write(s);
                    dataOutput.Flush();
                }
                return address;

            }
            else
            {

            }
            return -2;
        }

        public class PacketHandler
        {
            public Packet[] packetQueue;
            public Packet beatQueue;
            private Thread mainThread;
            public RRCPClient client;

            public PacketHandler(RRCPClient client)
            {
                this.packetQueue = new Packet[101]; //Makes packet queue
                this.beatQueue = null;
                this.client = client;
                this.mainThread = new Thread(run);
                this.mainThread.Start();
            }

            public void run()
            {
                while (client.isConnected())
                {
                    try
                    {
                        new Packet((sbyte)client.readNumber((sbyte)PacketTypes.Byte), client); //Makes new packet of id read from byte
                    }
                    catch (Exception ex)
                    {

                    }
                    Thread.Sleep(TIMEOUT_NUM);
                }
            }

            public Packet getPacket(sbyte address)
            {
                int i = 0;
                while (packetQueue[address] == null)
                {
                    ++i;
                    if (i > client.timeout + 10)
                    {
                        return null;
                    }
                    Thread.Sleep(TIMEOUT_NUM);
                }
                Packet p = packetQueue[address];
                packetQueue[address] = null;
                return p;
            }

            public Packet getHeartBeat()
            {
                int i = 0;
                while (beatQueue == null)
                {

                    if (i > client.timeout)
                    {

                        return new Packet((sbyte)100, client); //Returns error packet
                    }
                    Thread.Sleep(TIMEOUT_NUM);
                    ++i;
                }
                client.setHeartBeatDelay(i);
                Packet p = beatQueue;
                resetBeatQueue();
                return p;
            }

            public void addPacketToQueue(Packet p)
            {
                packetQueue[p.getAddress()] = p;
            }

            public void addPacketToBeatQueue(Packet p)
            {
                beatQueue = p;
            }

            public void resetBeatQueue()
            {
                beatQueue = null;
            }

        }

        public ValueType readNumber(sbyte id)
        {
            try
            {
                if (id == (sbyte)PacketTypes.Byte || id == (sbyte)PacketTypes.ByteArray) return dataInput.ReadSByte();
                else if (id == (sbyte)PacketTypes.Integer || id == (sbyte)PacketTypes.IntegerArray) return dataInput.ReadInt32();
                else if (id == (sbyte)PacketTypes.Double || id == (sbyte)PacketTypes.DoubleArray) {
                    byte[] b = dataInput.ReadBytes(8);
                    Array.Reverse(b);
                    return BitConverter.Int64BitsToDouble(BitConverter.ToInt64(b, 0));
                }
                else if (id == (sbyte)PacketTypes.Long || id == (sbyte)PacketTypes.LongArray) return dataInput.ReadInt64();
                else if (id == (sbyte)PacketTypes.Short || id == (sbyte)PacketTypes.ShortArray) return dataInput.ReadInt16();
                else if (id == (sbyte)PacketTypes.Float || id == (sbyte)PacketTypes.FloatArray) return dataInput.ReadSingle();
                else return -1;
            }
            catch (IOException ex)
            {

            }
            return -1;
        }

        public Object readNumberArray(sbyte id)
        {
            int length = (int)readNumber((sbyte)PacketTypes.Integer);
            Object[] n = new Object[length];
            for (int i = 0; i < length; i++)
            {
                n[i] = readNumber(id);
            }
            return n;
        }

        public bool readBoolean()
        {
            try
            {
                return dataInput.ReadBoolean();
            }
            catch (IOException ex)
            {

            }
            return false;
        }

        public string readString()
        {
            try
            {
                 byte[] b = dataInput.ReadBytes(2);
                 Array.Reverse(b);
                 Debug.WriteLine(new string(dataInput.ReadChars(BitConverter.ToUInt16(b, 0))));
                 return new string(dataInput.ReadChars(BitConverter.ToUInt16(b, 0)));
            }
            catch (IOException ex)
            {

            }
            return null;
        }

        public Object readPacket(sbyte address)
        {
            if (isConnected())
            {
                if (address == -2)
                {
                    return null;
                }
                Packet isNull = packetHandler.getPacket(address);
                if (isNull == null)
                {
                    return null;
                }
                return isNull.getData();
            }
            return null;
        }

        private void close()
        {
            this.connected = false;
            if (socket != null)
            {
                this.socket.Close();
            }
            if (autoReconnect)
            {
                this.Connect();
            }
        }

        private void HeartBeatRun()
        {
            while (isConnected() && sendHeartBeat())
            {
                Debug.WriteLine("Heart Beat");
                Thread.Sleep(250);
            }
            Debug.WriteLine("Heart Beat Dead!");
            close();
        }

        private bool sendHeartBeat()
        {
            lock (outputLock)
            {
                try
                {
                    Debug.WriteLine("Heart Beat Send");
                    dataOutput.Write((sbyte)21);
                    dataOutput.Flush(); 
                }
                catch (Exception e)
                {
                    return false;
                }
            }
            if (packetHandler.getHeartBeat().getID() == 21)
            {
                connected = true;
                return true;
            }
            else
            {
                return false;
            }
        }

        private void setHeartBeatDelay(int i)
        {
            this.heartBeatDelay = i;
        }

        public int getDelay()
        {
            return TIMEOUT_NUM * this.heartBeatDelay;
        }

        public class Packet
        {
            private sbyte id;
            private Object data;
            private sbyte address;

            public Packet(sbyte id, RRCPClient client)
            {
                Debug.WriteLine("Getting Something " + id);
                this.id = id;
                if (id == (sbyte)RRCPClient.PacketTypes.HeartBeat)
                { //Determains what type of packet it is from id 
                    client.packetHandler.addPacketToBeatQueue(this);
                }
                else if (id <= (sbyte)RRCPClient.PacketTypes.Float)
                { //Numbers
                    address = (sbyte)client.readNumber((sbyte)RRCPClient.PacketTypes.Byte);
                    data = client.readNumber(id);
                    client.packetHandler.addPacketToQueue(this);
                }
                else if (id == (sbyte)RRCPClient.PacketTypes.Boolean)
                {
                    address = (sbyte)client.readNumber((sbyte)RRCPClient.PacketTypes.Byte);
                    data = client.readBoolean();
                    client.packetHandler.addPacketToQueue(this);
                }
                else if (id == (sbyte)RRCPClient.PacketTypes.String)
                {
                    address = (sbyte)client.readNumber((sbyte)RRCPClient.PacketTypes.Byte);
                    data = client.readString();
                    client.packetHandler.addPacketToQueue(this);
                }
                else if (id >= (sbyte)RRCPClient.PacketTypes.DoubleArray && id <= (sbyte)RRCPClient.PacketTypes.FloatArray)
                {
                    address = (sbyte)client.readNumber((sbyte)RRCPClient.PacketTypes.Byte);
                    data = client.readNumberArray(id);
                    client.packetHandler.addPacketToQueue(this);
                }
                else if (id == 100)
                { //Error packet
                }
                else
                {
                    data = null;
                }
            }

            public Object getData()
            {
                return data;
            }

            public sbyte getID()
            {
                return id;
            }

            public sbyte getAddress()
            {
                return address;
            }
        }
    }

    
}
